messaging/src/main/java/io/lantern/messaging/metadata/Metadata.kt

```diff
@@ -176,61 +181,61 @@ class Metadata(
+            val targetBytesPerBar = Math.min(fileSize, maxBytesToProcess) / numberOfBars
             val codec = MediaCodec.createDecoderByType(mimeType)
             if (totalDurationUs == 0L) {
                 throw IOException("Zero duration")
             }
             codec.configure(format, null, null, 0)
             codec.start()
             extractor.selectTrack(0)
-            val kTimeOutUs: Long = 5000 // 5000 microseconds (5 milliseconds)
 
             audioDecoderExecutor.submit {
+                // read all encoded samples (mp3, opus, whatever) and submit them to the MediaCodec
+                // for decoding
                 var sawInputEOS = false
+                var scratchBuffer = ByteBuffer.allocate(1024768)
+                var bytesProcessedForCurrentBar = 0
                 while (!sawInputEOS) {
-                    val inputBufIndex = codec.dequeueInputBuffer(kTimeOutUs)
+                    val inputBufIndex = codec.dequeueInputBuffer(dequeueTimeoutUs)
                     if (inputBufIndex >= 0) {
                         val dstBuf = codec.getInputBuffer(inputBufIndex)!!
                         var sampleSize = extractor.readSampleData(dstBuf, 0)
-                        var presentationTimeUs: Long = 0
+                        bytesProcessedForCurrentBar += sampleSize
+                        var presentationTimeUs = 0L
                         if (sampleSize < 0) {
                             sawInputEOS = true
                             sampleSize = 0
                         } else {
                             presentationTimeUs = extractor.sampleTime
                         }
+
                         if (!sawInputEOS) {
-                            val barSampleIndex =
-                                (
-                                    SAMPLES_PER_BAR * (BAR_COUNT * extractor.sampleTime) /
-                                        totalDurationUs
-                                    ).toInt()
                             sawInputEOS = !extractor.advance()
-                            var nextBarSampleIndex =
-                                (
-                                    SAMPLES_PER_BAR * (BAR_COUNT * extractor.sampleTime) /
-                                        totalDurationUs
-                                    ).toInt()
-                            while (!sawInputEOS && nextBarSampleIndex == barSampleIndex) {
-                                sawInputEOS = !extractor.advance()
-                                if (!sawInputEOS) {
-                                    nextBarSampleIndex =
-                                        (
-                                            SAMPLES_PER_BAR * (BAR_COUNT * extractor.sampleTime) /
-                                                totalDurationUs
-                                            ).toInt()
+
+                            if (bytesProcessedForCurrentBar >= targetBytesPerBar) {
```

Comment:
This is the main part of the fix. For performance reasons, both the old and the new logic skip portions of the input to avoid decoding the whole file. On smaller files we don't skip anything, which is why waveform generation was more reliable on smaller files.

On bigger files, the old logic would skip based on a "sample index", but the size of samples is non-deterministic depending on what extractor.advance() does.

The new scheme skips based on number of bytes read, setting the cap not based on SAMPLES_PER_BAR but rather on the total number of bytes we're willing to process. This seems to work more reliably.